evm/src/fixed_recursive_verifier.rs

```diff
@@ -44,24 +47,47 @@ where
     F: RichField + Extendable<D>,
     C: GenericConfig<D, F = F>,
 {
-    /// The root circuit, which aggregates the (shrunk) per-table recursive proofs.
+    /// The EVM root circuit, which aggregates the (shrunk) per-table recursive proofs.
     pub root: RootCircuitData<F, C, D>,
+    pub aggregation: AggregationCircuitData<F, C, D>,
     /// Holds chains of circuits for each table and for each initial `degree_bits`.
     by_table: [RecursiveCircuitsForTable<F, C, D>; NUM_TABLES],
 }
 
-/// Data for the special root circuit, which is used to combine each STARK's shrunk wrapper proof
+/// Data for the EVM root circuit, which is used to combine each STARK's shrunk wrapper proof
 /// into a single proof.
 pub struct RootCircuitData<F, C, const D: usize>
 where
     F: RichField + Extendable<D>,
     C: GenericConfig<D, F = F>,
 {
-    pub circuit: CircuitData<F, C, D>,
+    circuit: CircuitData<F, C, D>,
     proof_with_pis: [ProofWithPublicInputsTarget<D>; NUM_TABLES],
     /// For each table, various inner circuits may be used depending on the initial table size.
     /// This target holds the index of the circuit (within `final_circuits()`) that was used.
     index_verifier_data: [Target; NUM_TABLES],
+    /// Public inputs used for cyclic verification. These aren't actually used for EVM root
+    /// proofs; the circuit has them just to match the structure of aggregation proofs.
+    cyclic_vk: VerifierCircuitTarget,
```

Comment:
nit: Could we remove this field and just have a generator set these targets to zeros (since they aren't actually used)?

Reply:
Hm I like the idea, just a minor complication... currently `conditionally_verify_cyclic_proof` expects its (inner) `cyclic_proof_with_pis`'s VK to match, even if `condition` is false so the cyclic proof wasn't selected for verification.

So in `prove_aggregation`, if an inner proof is an evm proof, we'll have an `agg_proof` target that's mostly ignored, except that its public VK must match. Currently we'll set both proof targets to the evm proof, which happens to have the agg proof's VK in its PIs, since `prove_root` sets it to that.

We could change `prove_aggregation` to always set the agg proof target's public VK, but I think it'd get a tad messy compared to the current code which can simply use `set_proof_with_pis_target`.

Maybe the most natural/elegant solution would be to change `conditionally_verify_cyclic_proof` so that it only enforces that the VKs match if `condition` is true? Though it'd cost a couple extra gates.

Reply:
I'll leave it for now but I can change this later depending on your feedback.

Reply:
Sounds good, definitely easier to keep it like that then 👍